presto-native-execution/presto_cpp/main/common/prometheus-metrics/PrometheusReporter.h

```diff
@@ -0,0 +1,110 @@
+ * limitations under the License.
+ */
+#include <prometheus/registry.h>
+#include "prometheus/collectable.h"
+#include "prometheus/counter.h"
+#include "prometheus/gauge.h"
+#include "prometheus/histogram.h"
+#include "prometheus/summary.h"
+#include "prometheus/text_serializer.h"
+#include "velox/common/base/Exceptions.h"
+#include "velox/velox/common/base/GTestMacros.h"
+#include "velox/velox/common/base/StatsReporter.h"
+
+namespace facebook::presto::prometheus {
+using CounterMap = folly::Synchronized<
+    std::unordered_map<std::string, ::prometheus::Counter&>>;
+using GaugeMap =
+    folly::Synchronized<std::unordered_map<std::string, ::prometheus::Gauge&>>;
+using HistogramMap = folly::Synchronized<
+    std::unordered_map<std::string, ::prometheus::Histogram&>>;
+using SummaryMap = folly::Synchronized<
+    std::unordered_map<std::string, ::prometheus::Summary&>>;
+
+static constexpr std::string_view kSummarySuffix("_summary");
+
+/*
+ * Prometheus CPP library exposes following classes:
+ * 1. Registry.
+ * 2. Family.
+ * 3. Specific metric type classes like Counter, Gauge, Histogram etc.
+ *
+ * A Family of metrics will have same metric name, for example, http_latency_ms.
+ * The different members of a family will have unique labels. For metric
+ * http_latency_ms labels could be {mehtod="GET"}, {method="PUT"},
+ * {method="POST"} etc. Prometheus treats {<metric_name>, [labels]} as unique
+ * metric object.
+ */
+class PrometheusReporter : public facebook::velox::BaseStatsReporter {
+ public:
+  explicit PrometheusReporter(const ::prometheus::Labels& labels)
+      : labels_(labels) {
+    registry_ = std::make_shared<::prometheus::Registry>();
+  }
+  void registerMetricExportType(const char* key, velox::StatType)
+      const override;
+
+  void registerMetricExportType(folly::StringPiece key, velox::StatType)
+      const override;
+
+  void registerHistogramMetricExportType(
+      const char* key,
+      int64_t bucketWidth,
+      int64_t min,
+      int64_t max,
+      const std::vector<int32_t>& pcts) const override;
+
+  void registerHistogramMetricExportType(
+      folly::StringPiece key,
+      int64_t bucketWidth,
+      int64_t min,
+      int64_t max,
+      const std::vector<int32_t>& pcts) const override;
+
+  void addMetricValue(const std::string& key, size_t value = 1) const override;
+
+  void addMetricValue(const char* key, size_t value = 1) const override;
+
+  void addMetricValue(folly::StringPiece key, size_t value = 1) const override;
+
+  void addHistogramMetricValue(const std::string& key, size_t value)
+      const override;
+
+  void addHistogramMetricValue(const char* key, size_t value) const override;
+
+  void addHistogramMetricValue(folly::StringPiece key, size_t value)
+      const override;
+
+  std::string getSerializedMetrics() const;
+
+ private:
+  // A map of labels assigned to each metric which helps in filtering at client
+  // end.
+  const ::prometheus::Labels labels_;
```

Comment:
Let's use PIMPL idiom and push this to the CPP file. 
See velox::S3FileSystem for example.
We don't want to expose Prometheus library headers into Prestissimo.